File: 2020/day2/part1/main_test.py
import unittest

from advent_of_code.day2.part1.main import Policy, make_policy, main

class TestDay2Part1(unittest.TestCase):

    def test_make_policy(self):
        input_data: str = "1-4 q: abcqrs"
        expected_output: Policy = Policy(lower_limit=1, upper_limit=4, character="q", password="abcqrs")
        self.assertEqual(make_policy(line=input_data).__dict__, expected_output.__dict__)

    def test_is_policy_valid(self):
        p1: Policy = Policy(*(1, 4, "q", "q"))
        self.assertTrue(p1.is_password_valid())
        p2: Policy = Policy(*(1, 4, "q", ""))
        self.assertFalse(p2.is_password_valid())
        p3: Policy = Policy(*(1, 4, "q", "qqqqq"))
        self.assertFalse(p3.is_password_valid())

    def test_main(self):
        input_data: int = main("tests/day2/input_test.txt")
        self.assertEqual(input_data, 2)
    

File: 2020/day2/part1/main_test.go
package part1

import (
	"advent-of-code/advent_of_code/day2/part1"
	"testing"
)

func TestMakePolicy(t *testing.T) {
	input := "1-4 q: abcqrs"
	expected := part1.Policy{1, 4, "q", "abcqrs"}
	calculated := part1.MakePolicy(input)

	if expected != *calculated {
		t.Errorf("Struct mismatch! Expected: %+v, got: %+v", expected, calculated)
	}
}

func TestIsPolicyValid(t *testing.T) {
	p1 := part1.Policy{1, 4, "q", "q"}
	if true != part1.IsPolicyValid(p1) {
		t.Errorf("Mismatch! Expected %t, got %t", true, part1.IsPolicyValid(p1))
	}
	p2 := part1.Policy{1, 4, "q", ""}
	if false != part1.IsPolicyValid(p2) {
		t.Errorf("Mismatch! Expected %t, got %t", false, part1.IsPolicyValid(p2))
	}
	p3 := part1.Policy{1, 4, "q", "qqqqq"}
	if false != part1.IsPolicyValid(p2) {
		t.Errorf("Mismatch! Expected %t, got %t", false, part1.IsPolicyValid(p3))
	}
}

func TestMain(t *testing.T) {
	input_data := part1.Main("../input_test.txt")
	if input_data != 2 {
		t.Errorf("Mismatch! Expected %d, got %d", 2, input_data)
	}
}
